Update accessibility tooling article (#39189)
mdn/content: files/en-us/learn_web_development/core/accessibility/tooling/index.md

## Accessibility tools

Let's have a look at the tools and techniques you can use for testing website accessibility and fixing the issues you uncover.

### Testing source order

> Your content should make logical sense in its source order — you can always place it where you want using CSS later on, but you should get the source order right to begin with. This is because assistive technologies tend to read out website content in source order, and disabled users often modify or turn off parts of the CSS to make the content more legible (common examples are increasing font size and applying high contrast color schemes).

Reviewer: Small reword for clarity if you like it:

```suggestion
Your content should make logical sense in its source order — you can always display it differently with CSS later on, but you should get the underlying structure correct to begin with. This is because assistive technologies read website content based on the order of the source, and disabled people often modify or turn off parts of the CSS to make content more legible (common examples are increasing font size and applying high contrast color schemes).
```

- Firefox: Select _View > Page Style > No Style_ from the main menu.
- Safari: [Open the browser developer tools](/en-US/docs/Learn_web_development/Howto/Tools_and_setup/What_are_browser_developer_tools#how_to_open_the_devtools_in_your_browser), click the _Device Settings_ button near the top-left of the developer tools panel (looks like a computer monitor), then check the "Disable CSS" checkbox in the panel that appears.
- Chrome/Edge: Install the [Web Developer Toolbar](https://chromewebstore.google.com/detail/web-developer/bfbameneiokkgbdmiekhjnmfkcnldhhm) extension, then restart the browser. Click the "Web Developer" gear icon that should now be available in your extensions menu, then select _CSS > Disable All Styles_.

### Color contrast checkers

When choosing a color scheme for your website, you should make sure that the text (foreground) color contrasts well with the background color. Your design might look cool, but it is no good if people can't read your content. Use a tool like WebAIM's [Color Contrast Checker](https://webaim.org/resources/contrastchecker/) to check whether your scheme is contrasting enough.

> Another tip is to not rely on color alone for signposts/information, as this will be no good for those with visual impairments like color blindness. Instead of marking required form fields in red, for example, mark them with an asterisk and in red.

Reviewer: ```suggestion
Another tip is to avoid using only color for signposts or highlighting important information, as this might be missed by people with visual impairments like color blindness. Instead of marking required form fields in red, for example, mark them with an asterisk and in red.
```

## Screen readers

> One of the most common types of assistive technology (AT) that disabled people use to consume web content — and one of the most common ones you'll use to test the accessibility of your webpages — is **screen readers**. These are pieces of software that read out webpage content — and content from other apps installed on the user's operating system — so that they can still use it content if they cannot see it.

Reviewer: Just breaking up the last sentence if you like it, you might have a better version

```suggestion
One of the most common types of assistive technology (AT) that disabled people use — and one of the most common ones you'll use to test the accessibility of your webpages — is **screen readers**. These are pieces of software that read out webpage content or content from other apps installed on someone's operating system. Screen readers enable people to use the web and other applications without having to see any visual content.
```

Author: Used with a tweak

Web browsers expose information about the page's content for screen readers (and other AT) to communicate to the user through a representation called the [accessibility tree](/en-US/docs/Glossary/Accessibility_tree). This provides semantic information such as names and descriptions of elements, what their purpose or role is (is it a button, or an input field?), and whether they are in aparticular state (for example, is a dialog box open or closed?)

> This information is pretty trivial in the case of a paragraph of text, which sounds pretty much how it is written, but it becomes really important when describing complex widgets to a visually impaired user, such as a drop-down menu or a video player. This is also why it is very important to use semantic HTML elements correctly, which you'll look at in detail in the next article in this module. If you mark up content using the wrong element, it can really confuse screen reader users.

Reviewer: ```suggestion
This information might be trivial in the case of a paragraph of text, which sounds pretty much how it is written, but it can get complicated when it comes to user interfaces, such as a drop-down menu or a video player. This is why it's very important to use semantic HTML correctly, which you'll look at in detail in the next article in this module. If you mark up content using the wrong element, it can confuse screen reader users.
```

Author: Used, but with a tweak

> It is definitely worth making sure you have a screen reader or two installed on your development machine, and try using your favourite websites via a screenreader, as discussed below. Understanding how severely visually impaired people use the web is key to designing products that work better for them.

Reviewer: ```suggestion
Make sure you have a screen reader or two installed on your development machine, and try using your favorite websites via a screen reader, as discussed below. Understanding how visually impaired people use the web is key to designing products that work better for everyone.
```

> ### What screenreaders are available?

Reviewer: Should we use this everywhere?

```suggestion
### What screen readers are available?
```

There are several screen readers available:

- Some are paid-for commercial products, like [JAWS](https://www.freedomscientific.com/Products/software/JAWS/) (Windows).
- Some are free products, like [NVDA](https://www.nvaccess.org/) (Windows), [ChromeVox](https://support.google.com/chromebook/answer/7031755) (Chrome, Windows, and macOS), and [Orca](https://wiki.gnome.org/Projects/Orca) (Linux).
- Some are built into the operating system, like [VoiceOver](https://www.apple.com/accessibility/vision/) (macOS and iOS), [ChromeVox](https://support.google.com/chromebook/answer/7031755) (on Chromebooks), and [TalkBack](https://play.google.com/store/apps/details?id=com.google.android.marvin.talkback) (Android).

> Generally, screen readers are separate apps that run on the host operating system and can read not only web pages, but content in other apps as well (this is not always the case; ChromeVox for example is a browser extension). Screen readers tend to have some differences in exact behavior and controls, so you'll have to consult the documentation for your chosen screen reader to get all the details. Saying that, they all work in basically the same sort of way.

Reviewer: ```suggestion
Generally, screen readers are separate apps that run on the host operating system and can read web pages and content in other apps as well (this is not always the case; ChromeVox for example is a browser extension). Screen readers tend to have some differences in exact behavior and controls, so you'll have to consult the documentation for your chosen screen reader to get all the details. Saying that, they all work in basically the same sort of way.
```

- Joysticks: A control stick that can be moved to move the pointer around. Joysticks are less precise than trackballs but usable by people with a wide range of physical impairments, even severe disabilities.
- Touchpads: Most modern laptops have a touchpad (sometimes called a trackpad) — a flat tactile sensor allowing you to move the pointer around with a finger, as well as performing multi-finger gestures in the same way as mobile gestures. You can buy external touchpads for devices that don't have internal ones. Some people find them to be more precise than mice.

### Screen magnifiers

Screen magnifiers provide partially-sighted users with a magnified view of their device's display to enable them to understand and interact with device content more easily, as well as providing other features such as color adjustment to help with color blindness, and adjusting the size of mouse pointers and text cursors to make them easier to see.

Software and hardware screen magnifiers are available:

> - Most modern operating systems have a built-in app for magnifying all or part of the screen, for example Zoom on mac or Magnifier on Windows. They also tend to provide options for univerally increasing the text size, mouse cursor size, etc. Third-party options are available as well.

Reviewer: ```suggestion
- Most modern operating systems have a built-in app for magnifying all or part of the screen, for example Zoom on mac or Magnifier on Windows. They also tend to provide options for universally increasing the text size, mouse cursor size, etc. Third-party options are available as well.
```

- Hardware screen magnifiers tend to consist of a separate screen that sits next to or in front of your device's screen and projects a larger version of it, or a zoomed version of part of it.

### Voice recognition software

> Voice (or speech) recognition software allows you to speak commands to control your device and/or speak the text of emails or documents and have the computer write the text for you. This is very useful for people with who are unable to use a keyboard or other control mechanisms.

Reviewer: ```suggestion
Voice (or speech) recognition software allows you to speak commands to control your device and/or speak the text of emails or documents and have the computer write the text for you. This is very useful for people who are unable to use a keyboard or other control mechanisms.
```

> Modern operating systems have features built in to enable this (for example Dictation on mac, or Voice Access on Windows), and there are third party apps available too, from desktop apps to browser extensions.

Reviewer: ```suggestion
Modern operating systems have features built in to enable this (for example Dictation on mac, or Voice Access on Windows), and there are third-party apps available too, from desktop apps to browser extensions.
```
